dist/common_functions.sh

```diff
@@ -0,0 +1,18 @@
+#!/bin/sh
+
+# Common shell functions shared across dist/ scripts
+
+# Find the commit from develop at which point the current branch diverged.
+# rev-list will show all commits that are present on our current branch but not on develop, and 
+# the oldest of these is our first commit post-divergence. If this commit exists then 
+# we can take its parent. If no such commits exist then we're currently on a commit in the 
+# develop branch and can use HEAD instead
+last_commit_from_dev() {
```

Comment:
Suggestion: we could have Python call this or bash call Python instead of having "two" implementations.

Reply:
I don't mind this but I can't see a clean way to make the change. If we called bash code from python we'd need to append `. ./common_functions.py` to every subfunction call, and if we called python from bash we'd need to define a __main__ function that can call each of the functions defined within.
Open to any other implementation suggestions if you have them

Reply:
```
and if we called python from bash we'd need to define a main function that can call each of the functions defined within
```
I think that could be totally fine!

Reply:
We can do the following with a touch of reflection:
```
if __name__ == "__main__":
    # Allow users to execute any function defined in this file via the command line
    # If a function name is provided and that function exists in this file then execute 
    # it along with any provided arguments.
    functions_in_file = inspect.getmembers(sys.modules[__name__], predicate=inspect.isfunction)
    functions = dict(functions_in_file)

    if len(sys.argv) > 1:
        name, *args = sys.argv[1:]

        if name in functions:
            print(functions[name](*args))
            exit(0)
        else:
            print(f"Function name '{name}' not recognised!")
    else:
        print("Usage: `python3 ./common_functions.py function_name [function_args...]`")

    exit(1)
```
allowing us to call it from bash with `python3 ./common_functions.py last_commit_from_dev`.

I'll think on this a bit.. It's not my favourite solution but it'll do away with the redundant definitions of `last_commit_from_dev`